addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/ThirdGeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,767 @@
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * The {@link ThirdGenerationHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author René Stakemeier - Initial contribution
+ */
+@NonNullByDefault
+public class ThirdGenerationHandler extends BaseThingHandler {
+
+    private ThirdGenerationInverterTypes inverterType;
+
+    // GSON handler
+    private static final Gson GSON = new GsonBuilder().setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
+    // used for logging
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    // base URL of the web api
+    private static final String WEB_API = "/api/v1";
+
+    /*
+     * operations used for authentication
+     */
+    private static final String AUTH_START = "/auth/start";
+    private static final String AUTH_FINISH = "/auth/finish";
+    private static final String AUTH_CREATE_SESSION = "/auth/create_session";
+
+    /*
+     * operations used for gathering process data from the device
+     */
+    private static final String PROCESSDATA = "/processdata";
+
+    /*
+     * List of available channels for the concrete device
+     */
+
+    /*
+     * After the authentication the result (the session id) is stored here and used to "sign" future requests
+     */
+    @Nullable
+    protected String sessionId = null;
+    /*
+     * The configuration file containing the host, the password and the refresh interval
+     */
+    @Nullable
+    private ThirdGenerationConfiguration config;
+
+    @Nullable
+    protected ScheduledFuture<?> refreshScheduler;
+
+    /**
+     * Constructor of this class
+     *
+     * @param thing the thing
+     */
+    public ThirdGenerationHandler(Thing thing, ThirdGenerationInverterTypes inverterType) {
+        super(thing);
+        this.inverterType = inverterType;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        /*
+         * All channels are readonly and updated by the scheduler
+         */
+    }
+
+    @Override
+    public void dispose() {
+        if (refreshScheduler != null) {
+            refreshScheduler.cancel(true);
+        }
+        super.dispose();
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public void initialize() {
+        config = getConfigAs(ThirdGenerationConfiguration.class);
+        /*
+         * temporary value while initializing
+         */
+        updateStatus(ThingStatus.UNKNOWN);
+
+        /*
+         * Start the initialization
+         */
+        authenticate();
```

Comment:
You should not call this method directly in `initialize`. `initialize` should finish quicky. Better schedule the call. If refreshScheduler doesn't make sense yet before being authenticated then start scheduler in `authenticate`.

Reply:
It´s now executed by the scheduler